README.md
# eos-application-tools

# Description

This package currently contains one component:

  * eos-install-app-helper: Wrapper application to run apps that
    we recommend but cannot distribute with Endless images.

## eos-install-app-helper

This package provides system level wrapper applications to allow easily
downloading, installing and running specific apps on Endless OS.

This wrapper application does mainly two things when you click on the desktop icon:

  * If the app is not yet installed, the wrapper script opens the app's detail
    page in the App Center (GNOME Software).

  * If the app has already been installed (using flatpak as a delivery mechanism),
    the wrapper script launches the app.

All these files will be installed, exceptionally, as part of the OSTree, so that the
icon and the wrapper app are available on the desktop at any time, either to run
the app or to install it if not yet available.

## License

eos-application-tools is Copyright (C) 2016, 2017 Endless Mobile, Inc.
and is licensed under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2 of
the License, or (at your option) any later version.

See the COPYING file for the full version of the GNU GPLv2 license


debian/changelog
eos-application-tools (1.0.4) eos; urgency=medium

  * Split Flatpak auto-installation JSON files to architecture-independent
    eos-flatpak-autoinstall package so they are included on ARM systems.

  https://phabricator.endlessm.com/T22652

 -- Robert McQueen <rob@endlessm.com>  Mon, 21 May 2018 16:53:47 +0100

eos-application-tools (1.0.3) eos; urgency=medium

  * Remove eos-browser-helper and Google Chrome specifics from this package,
    so that eos-browser-tools can be reinstated.

 -- Robert McQueen <rob@endlessm.com>  Fri, 04 Aug 2017 09:37:56 +0100

eos-application-tools (1.0.2) eos; urgency=medium

  * Renamed to eos-application-tools, because the package has been made more
    generic.

 -- Cosimo Cecchi <cosimo@endlessm.com>  Wed, 02 Aug 2017 16:07:31 +0100

eos-browser-tools (1.0.1) eos; urgency=medium

  * Renamed to eos-browser-tools, so that we can ship browser related
    tools separately from the eos-google-chrome-helper package.

  https://phabricator.endlessm.com/T15917

 -- Mario Sanchez Prada <mario@endlessm.com>  Tue, 07 Mar 2017 11:42:24 +0000

eos-google-chrome-helper (1.0.0) eos; urgency=medium

  * Initial release

 -- Mario Sanchez Prada <mario@endlessm.com>  Thu, 10 Nov 2016 13:20:12 +0000


src/eos-install-app-helper-installer.py
#!/usr/bin/python3
#
# eos-install-app-helper-installer: helper script to install an App
#
# Copyright (C) 2016, 2017 Endless Mobile, Inc.
# Authors:
#  Michal Rostecki <michal@kinvolk.io>
#  Mario Sanchez Prada <mario@endlessm.com>
#  Sam Spilsbury <sam@endlessm.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import argparse
import logging
import subprocess
import sys

import gi
gi.require_version('Flatpak', '1.0')
from gi.repository import Flatpak
from gi.repository import Gio
from gi.repository import GLib
from systemd import journal


def exit_with_error(*args):
    logging.error(*args)
    sys.exit(1)


class InstallAppHelperInstaller:
    def __init__(self,
                 app_id,
                 remote,
                 branch):
        try:
            self._installation = Flatpak.Installation.new_system()
        except GLib.Error as e:
            exit_with_error("Couldn't not find current system installation: %r", e)

        if self._check_app_flatpak_launcher(app_id):
            logging.info("{app_id} is already installed".format(app_id))
            return

        logging.info("Could not find flatpak launcher for {}.".format(app_id))

        self._run_app_center_for_app(app_id, remote, branch)


    def _check_app_flatpak_launcher(self, app_id):
        try:
            self._installation.get_current_installed_app(app_id, None)
        except GLib.Error as e:
            logging.info("{} application is not installed".format(app_id))
            return False
        return True

    def _wait_for_installation(self, app_id):
        def _installation_finished(monitor, file_, other_file, event_type):
            if event_type != Gio.FileMonitorEvent.CHANGES_DONE_HINT:
                return

            if self._check_app_flatpak_launcher(app_id):
                logging.info("{} has been installed".format(app_id))
                loop.quit()

        loop = GLib.MainLoop()

        monitor = self._installation.create_monitor(None)
        monitor.connect('changed', _installation_finished)

        loop.run()

    def _get_unique_id(self, app_id, remote_name, branch):
        app_app_center_id = app_id

        try:
            remote = self._installation.get_remote_by_name(remote_name)
        except GLib.Error as e:
            logging.warning("Could not find flatpak remote {}: {}".format(remote, str(e)))

        # Get the default branch if no branch has been given, as it's needed to construct
        # the full unique ID GNOME Software expects.
        if not branch:
            branch = remote.get_default_branch()

        if branch:
            app_app_center_id = 'system/flatpak/{}/desktop/{}.desktop/{}'.format(remote_name,
                                                                                 app_id,
                                                                                 branch)
        return app_app_center_id

    def _run_app_center_for_app(self, app_id, remote, branch):
        # FIXME: Ideally, we should be able to pass the app ID to GNOME Software
        # and it would do the right thing by opening the page for the app's branch matching
        # the default branch for the apps' source remote. Unfortunately, this is not the case
        # at the moment and fixing it is non-trivial, so we'll construct the full unique ID
        # that GNOME Software expects, right from here, based on the remote's metadata.
        unique_id = self._get_unique_id(app_id, remote, branch)

        logging.info("Opening App Center for {}...".format(unique_id))
        app_center_argv = ['gnome-software', '--details={}'.format(unique_id)]

        try:
            subprocess.Popen(app_center_argv)
        except OSError as e:
            exit_with_error("Could not launch {}: {}".format(app_id, repr(e)))

        self._wait_for_installation(app_id)

        if not self._check_app_flatpak_launcher(app_id):
            exit_with_error("{} isn't installed - something went wrong in GNOME Software".format(app_id))

        logging.info("{} successfully installed".format(app_id))


def main():
    # Send logging messages both to the console and the journal
    logging.basicConfig(level=logging.INFO)
    logging.root.addHandler(journal.JournalHandler())

    parser = argparse.ArgumentParser()
    parser.add_argument('--debug', dest='debug', action='store_true')
    parser.add_argument('--app-id', dest='app_id', help='Flatpak App ID', type=str, required=True)
    parser.add_argument('--remote', dest='remote', help='Flatpak Remote', type=str, required=True)
    parser.add_argument('--branch', dest='branch', help='Flatpak Branch', type=str, default='', required=False)
    parser.add_argument('--required-archs', dest='required_archs', default=[], nargs='*', type=str)

    parsed_args = parser.parse_args()

    if parsed_args.debug:
        logging.root.setLevel(logging.DEBUG)

    if parsed_args.required_archs and Flatpak.get_default_arch() not in parsed_args.required_archs:
        exit_with_error("Found installation of unsupported architecture: {}".format(parsed_args.required_archs))

    InstallAppHelperInstaller(parsed_args.app_id,
                              parsed_args.remote,
                              parsed_args.branch)
    sys.exit(0)


if __name__ == '__main__':
    main()


install-app-helper/eos-install-app-system-helper.py
#!/usr/bin/python3
#
# eos-install-app-system-helper: system helper to touch / reset the stamp file
#
# Copyright (C) 2017 Endless Mobile, Inc.
# Authors:
#  Mario Sanchez Prada <mario@endlessm.com>
#  Sam Spilsbury <sam@endlessm.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import argparse
import config
import logging
import os
import sys

from systemd import journal


def exit_with_error(*args):
    logging.error(*args)
    sys.exit(1)


def touch_file(file):
    with open(file, 'a'):
        os.utime(file, None)


def create_stamp_file(filename):
    dirname = os.path.dirname(filename)

    try:
        os.makedirs(dirname, exist_ok=True)
        touch_file(filename)
        logging.info("Stamp file created at {}".format(filename))
    except OSError as e:
        exit_with_error("Error creating stamp file at {}: {}"
                        .format(filename, str(e)))


def remove_stamp_file(filename):
    try:
        os.unlink(filename)
        logging.info("Stamp file removed from {}".format(filename))
    except OSError as e:
        exit_with_error("Error removing stamp file at {}: {}"
                        .format(filename, str(e)))


if __name__ == '__main__':
    # Send logging messages both to the console and the journal
    logging.basicConfig(level=logging.INFO)
    logging.root.addHandler(journal.JournalHandler())

    parser = argparse.ArgumentParser(description="Marks application installation as installed, system-wide")

    parser.add_argument('--debug', help="Show extra messages", action='store_true')
    parser.add_argument('--reset', help="Remove stamp file", action='store_true')
    parser.add_argument('--app-id', help="Application ID to stamp", type=str, required=True)

    parsed_args = parser.parse_args()
    if parsed_args.debug:
        logging.root.setLevel(logging.DEBUG)

    if parsed_args.reset:
        remove_stamp_file(config.STAMP_FILE_INITIAL_SETUP_DONE.format(app_id=parsed_args.app_id))
    else:
        create_stamp_file(config.STAMP_FILE_INITIAL_SETUP_DONE.format(app_id=parsed_args.app_id))

    sys.exit(0)


src/eos-install-app-helper.py
#!/usr/bin/python3
#
# eos-install-app-helper: helper script to install/launch an App.
#
# Copyright (C) 2016, 2017 Endless Mobile, Inc.
# Authors:
#  Mario Sanchez Prada <mario@endlessm.com>
#  Sam Spilsbury <sam@endlessm.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import argparse
import logging
import os
import subprocess
import sys

import config
import gi
gi.require_version('Flatpak', '1.0')
from gi.repository import Flatpak
from gi.repository import Gio
from gi.repository import GLib
from systemd import journal


def exit_with_error(message):
    logging.error(message)
    sys.exit(1)


class InstallAppHelperLauncher:
    def __init__(self,
                 app_id,
                 remote,
                 branch,
                 params):
        self._params = params
        try:
            self._installation = Flatpak.Installation.new_system()
        except GLib.Error as e:
            exit_with_error("Could not find current system installation: {}".format(repr(e)))

        self._start(app_id, remote, branch)

    def _start(self, app_id, remote, branch):
        is_installed = self._is_flatpak_installed(app_id)
        if is_installed:
            logging.info("Flatpak for {} found. Launching...".format(app_id))
            self._run_app(app_id, self._params)
        else:
            logging.info("Could not find flatpak for {}. Running installation script...".format(app_id))
            self._install_app_id(app_id,
                                 remote,
                                 branch)

    def _run_app(self, app_id, params):
        logging.info("Launching {} flatpak app through desktop file".format(app_id))
        desktop_id = app_id + '.desktop'
        desktop_info = Gio.DesktopAppInfo.new(desktop_id)
        if desktop_info:
            try:
                desktop_info.launch()
            except GLib.Error as e:
                exit_with_error("Could not launch {}: {}".format(app_id, repr(e)))
        else:
           exit_with_error("Could not find desktop file for {}".format(app_id))

    def _install_app_id(self,
                        app_id,
                        remote,
                        branch):
        extra_args = []
        if branch:
            extra_args.extend(['--branch', branch])
        try:
            subprocess.Popen([os.path.join(config.PKG_DATADIR, 'eos-install-app-helper-installer.py'),
                              '--app-id', app_id,
                              '--remote', remote] + extra_args)
        except OSError as e:
            exit_with_error("Could not launch {}: {}".format(app_id, repr(e)))

    def _is_flatpak_installed(self, app_id):
        try:
            self._installation.get_current_installed_app(app_id, None)
            return True
        except GLib.Error:
            logging.info("{} application is not installed".format(app_id))
            return False

    def _get_app_flatpak_launcher(self, app_id, launcher_name):
        app = None
        try:
            app = self._installation.get_current_installed_app(app_id, None)
        except GLib.Error:
            logging.info("{} application is not installed".format(app_id))
            return None

        app_path = app.get_deploy_dir()
        if not app_path or not os.path.exists(app_path):
            exit_with_error("Could not find {}'s application directory".format(app_id))

        app_launcher_path = os.path.join(app_path, 'files', 'bin', launcher_name)
        if not os.path.exists(app_launcher_path):
            exit_with_error("Could not find flatpak launcher for {}".format(app_id))

        logging.info("Found flatpak launcher for {}: %{}".format(app_id, repr(app_launcher_path)))
        return app_launcher_path


if __name__ == '__main__':
    # Send logging messages both to the console and the journal
    logging.basicConfig(level=logging.INFO)
    logging.root.addHandler(journal.JournalHandler())

    parser = argparse.ArgumentParser()
    parser.add_argument('--debug', dest='debug', action='store_true')
    parser.add_argument('--app-id', dest='app_id', help='Flatpak App ID', type=str, required=True)
    parser.add_argument('--remote', dest='remote', help='Flatpak Remote', type=str, required=True)
    parser.add_argument('--branch', dest='branch', help='Flatpak Branch', type=str, default='', required=False)
    parser.add_argument('--required-archs', dest='required_archs', default=[], nargs='*', type=str)

    parsed_args, otherargs = parser.parse_known_args()

    if parsed_args.debug:
        logging.root.setLevel(logging.DEBUG)

    # Some apps are only available for certain architectures
    if parsed_args.required_archs and Flatpak.get_default_arch() not in parsed_args.required_archs:
        exit_with_error("Found installation of unsupported architecture: {}".format(parsed_args.required_archs))

    InstallAppHelperLauncher(parsed_args.app_id,
                             parsed_args.remote,
                             parsed_args.branch,
                             otherargs)
    sys.exit(0)
